src/extensions/dnatco/confal-pyramids/util.ts

```diff
@@ -296,4 +310,95 @@ export namespace ConfalPyramidsUtil {
+                const resno = label_seq_id.value(residueIndex[unitElements[uI]]);
+                const altId = label_alt_id.value(unitElements[uI]);
+                if (resno === resnoLast) {
+                    uISecond = uI;
+                    break;
+                } else {
+                    if (altId !== '' && !possibleFirstAltIds.includes(altId)) possibleFirstAltIds.push(altId);
+                }
+            }
+            if (i === NI)
+                return void 0; // No valid second residue
+
+            // Fill possible second residue alt ids array
+            for (i; i < NI; i++) {
+                const uI = OrderedSet.getAt(indices, i);
+                const altId = label_alt_id.value(unitElements[uI]);
+                if (altId !== '' && !possibleSecondAltIds.includes(altId)) possibleSecondAltIds.push(altId);
+            }
+
+            if (possibleFirstAltIds.length > 1 || possibleSecondAltIds.length > 1) return void 0; // We got selection with more than one conformation
+
+            const firstResidue = { index: residueIndex[unitElements[uIFirst]], start: uIFirst, end: uISecond - 1 };
+            const secondResidue = { index: residueIndex[unitElements[uISecond]], start: uISecond, end: uILast };
+
+            try {
+                const { firstAtoms, secondAtoms } = this.residuesToAtoms(firstResidue, possibleFirstAltIds, secondResidue, possibleSecondAltIds);
+
+                const modelNum = this.hasMultipleModels ? this.modelNum : -1;
+                const firstLoc = StructureElement.Location.create(this.structure, this.unit, -1 as ElementIndex);
+                const secondLoc = StructureElement.Location.create(this.structure, this.unit, -1 as ElementIndex);
+                for (let i = 0; i < firstAtoms.length; i++) {
+                    const first = firstAtoms[i];
+                    for (let j = 0; j < secondAtoms.length; j++) {
+                        const second = secondAtoms[j];
+                        firstLoc.element = first;
+                        secondLoc.element = second;
+
+                        const name = this.stepToName(this.entryId, modelNum, firstLoc, secondLoc, '', '');
+                        const { pyramid } = this.getPyramidByName(name);
+                        if (pyramid !== undefined) return pyramid;
+                    }
+                }
+
+                return void 0;
+            } catch (e) {
+                return void 0;
+            }
+        }
+
+        constructor(unit: Unit.Atomic, private loci: StructureElement.Loci) {
+            super(loci.structure, unit);
+        }
+    }
+
+    export function lociToPyramid(loci: Loci) {
+        if (loci.kind !== 'element-loci') return void 0;
+        if (loci.elements.length < 1) return void 0;
+        if (!DC.isApplicable(loci.structure.model)) return void 0;
+        if (!Unit.isAtomic(loci.elements[0].unit)) return void 0;
+
+        return (new PyramidGetter(loci.elements[0].unit, loci)).get();
```

Comment:
This seems terribly inefficient, triggering very non-trivial computation each time a property is accessed.

A better solution is to create a ``FormatPropertyProvider`` which gets assigned the first time the user tries to access it which contains some index computed from the ``ndb_struct_ntc_*`` categories.

Reply:
I took a look at `FormatPropertyProvider` and I am not sure how would it help me in this particular situation.

In order to fetch a pyramid the code has to check if a given `Loci` represents a valid NtC step. Since this check allows me to get `ResidueIndex` of both residues I suppose a could create a map with indices of both residues as key and map them to pyramids. Do you have something this in mind?

Reply:
Well, you scan thru the entire tables in the categories. You could at least create an index on the tables so you don't have to effectively parse the data each time?

Reply:
Since a step maps to two residues I could create a `ResidueIndex` to `Pyramid` mapping right away in `createPyramidfFromCif()`, get the residue indices from `Loci` and just fetch the pyramid from the map. Would this be an acceptable solution?

Reply:
Yes, you can create the map when you create the pyramid representation and store it using the ``FormatPropertyProvider`` on the model. That way the pyramid label access would be instant.

Reply:
If this idea turns out as workable (proper NtC steps mapping can get pretty awkward with weird structures) it should allow me to eliminate most of the clumsy lookup code from util. I will give it a try and see where it breaks :)

Reply:
I considered the possibility of organising the NtC data in a way that would allow for more efficient lookups. Unfortunately, it is not that straightforward. I will try to explain what the problem is...

Handling of NtC steps is awkward because of alternate positions. While altpos is an atom propery, NtC steps need to "inflict" the fact that there are some atoms with altposes onto the entire residue. If I were to use `ResidueIndex` (actually two of them, one for each residue in the step) as a key in a map, such a key could possibly map to multiple steps and it would not be appararent which step to take without some extra logic.

When the current code gets two residues that may form a step, it has to:
- Check if there are any atoms with altpos on any of the residues
- Create a list of all possible step names, considering the altpos
- Check if any of the possible step names is present in the structure
- Return a pyramid for the found step

While this is not particularly nice it seems to be the only reliable way how to handle this. NtC pyramids are stored in an array and there is a "step name" -> "array index" map to allow for at least somewhat efficient access.

When the `PyramidGetter.get()` gets a `Loci`, it needs to:
- Check if the contained elements form two consecutive residues
- Walk through the elements of both residues to look for altposes
- Check if the contained elements actually form a step and return the corresponding pyramid if true

Given the explanation above, is there still something I should be looking to optimise?

Reply:
Can you share a file that has the ``ndb_struct_ntc_*`` categories?

Reply:
Also, should the ``lociToPyramid`` contain a check if the ``ConfalPyramidsProvider`` is assigned?

Also ``if (!Unit.isAtomic(loci.elements[0].unit)) return void 0;`` this check is cheaper than the one above it, so should move it there.

Reply:
There you go: Each of these structures contains at least a few NtC steps with altposes

https://blackbox.ibt.biocev.org/devel/RCSB/cif_dnatco_updated/3vok_v32C35A23.cif.gz
https://blackbox.ibt.biocev.org/devel/RCSB/cif_dnatco_updated/1dk1_v32C35A23.cif.gz
https://blackbox.ibt.biocev.org/devel/RCSB/cif_dnatco_updated/6r93_v32C35A23.cif.gz
https://blackbox.ibt.biocev.org/devel/RCSB/cif_dnatco_updated/3ok4_v32C35A23.cif.gz